src/python/pants/engine/legacy/graph.py

```diff
@@ -269,6 +275,93 @@ def _inject_specs(self, subjects):
+      all_valid_addresses.add(target_adaptor.address)
+    inst._validate(all_valid_addresses)
+    return inst
+
+  def __init__(self, target_types, address_mapper):
+    # TODO: Dependencies and implicit dependencies are mapped independently, because the latter
+    # cannot be validated until:
+    #  1) Subsystems are computed in engine: #5869. Currently instantiating a subsystem to find
+    #     its injectable specs would require options parsing.
+    #  2) Targets-class Subsystem deps can be expanded in-engine (similar to Fields): #4535,
+    self._dependent_address_map = defaultdict(set)
+    self._implicit_dependent_address_map = defaultdict(set)
+    self._target_types = target_types
+    self._address_mapper = address_mapper
+
+  def _validate(self, all_valid_addresses):
+    """Validate that all of the dependencies in the graph exist in the given addresses set."""
+    for dependency, dependents in iteritems(self._dependent_address_map):
+      if dependency not in all_valid_addresses:
+        raise AddressLookupError(
+            'Dependent graph construction failed: {} did not exist. Was depended on by:\n  {}'.format(
+             dependency.spec,
+             '\n  '.join(d.spec for d in dependents)
+          )
+        )
+
+  def _inject_target(self, target_adaptor):
+    """Inject a target, respecting all sources of dependencies."""
+    target_cls = self._target_types[target_adaptor.type_alias]
+
+    declared_deps = target_adaptor.dependencies
+    implicit_deps = (Address.parse(s,
+                                   relative_to=target_adaptor.address.spec_path,
+                                   subproject_roots=self._address_mapper.subproject_roots)
+                     for s in target_cls.compute_dependency_specs(kwargs=target_adaptor.kwargs()))
+    for dep in declared_deps:
+      self._dependent_address_map[dep].add(target_adaptor.address)
+    for dep in implicit_deps:
+      self._implicit_dependent_address_map[dep].add(target_adaptor.address)
+
+  def dependents_of_addresses(self, addresses):
+    """Given an iterable of addresses, yield all of those addresses dependents."""
+    seen = set(addresses)
+    for address in addresses:
+      seen.update(self._dependent_address_map[address])
+      seen.update(self._implicit_dependent_address_map[address])
+    return seen
+
+  def transitive_dependents_of_addresses(self, addresses):
+    """Given an iterable of addresses, yield all of those addresses dependents, transitively."""
+    addresses_to_visit = set(addresses)
+    while 1:
+      dependents = self.dependents_of_addresses(addresses)
+      # If we've exhausted all dependencies or visited all remaining nodes, break.
+      if (not dependents) or dependents.issubset(addresses_to_visit):
+        break
+      addresses = dependents.difference(addresses_to_visit)
+      addresses_to_visit.update(dependents)
+
+    return self.dependents_of_addresses(addresses_to_visit)
```

Comment:
This last call should always no-op, right? Skip it and just return `addresses_to_visit`?